### PR TITLE
feat(manager/cargo): support private registry index env var

### PR DESCRIPTION
**Pull request by tofay**
## Changes

Cargo supports defining private registry index URLs in environment variables rather than `.cargo/config[.toml]`. This PR adds support for determining registry URLs from the same variable. 

## Context

I have many Rust projects that use a private registry. We define this in Cargo's home directory of our build environment, rather than in a project-local file. I'm running renovate self-hosted.

This PR enables Renovate to discover the registry URL without adding the cargo configuration file to all projects.

## Documentation (please check one with an [x])

- [x] I have updated the documentation, or
- [ ] No documentation update is required

## How I've tested my work (please select one)

I have verified these changes via:

- [ ] Code inspection only, or
- [ ] Newly added/modified unit tests, or
- [ ] No unit tests but ran on a real repository, or
- [x] Both unit tests + ran on a real repository

I've been running this self-hosted against multiple private repositories and renovate has successfully determined the registry URLs.

### CODE REVIEW

File: lib/modules/manager/cargo/extract.spec.ts [hunk 1]
```diff
@@ -113,6 +116,18 @@ describe('modules/manager/cargo/extract', () => {
...
 
+    it('extracts registry urls from environment', async () => {
+      process.env.CARGO_REGISTRIES_PRIVATE_CRATES_INDEX =
+        'https://dl.cloudsmith.io/basic/my-org/my-repo/cargo/index.git';
+      process.env.CARGO_REGISTRIES_MCORBIN_INDEX =
+        'https://github.com/mcorbin/testregistry';
+      const res = await extractPackageFile(cargo6toml, 'Cargo.toml', {
+        ...config,
+      });
+      expect(res?.deps).toMatchSnapshot();
```
Comment (viceice) [inline lib/modules/manager/cargo/extract.spec.ts:127]: no snapshots, read our dev best practices guide

**Review by viceice — COMMENTED** [review]
you can already override the registry url via renovate config, so I think we shouldn't implement it this way.

instead renovate should convert the registry url to the ENV value before calling cargo to update the lockfile.

**Comment by tofay** [comment]
> you can already override the registry url via renovate config, so I think we shouldn't implement it this way.

I'm not trying to override the default registry URL here, but provide a URL for an additional index.
Renovate also needs to associate the registry name as specified in Cargo.toml with the index URL - I can't see how to that from the docs but if it's possible then I agree this PR can be discarded (https://docs.renovatebot.com/rust/ doesn't mention a `registryUrls` field unlike  other managers e.g pip/maven)

File: docs/usage/rust.md [hunk 2]
```diff
@@ -11,14 +11,19 @@ Renovate supports upgrading dependencies in `Cargo.toml` files and their accompa
 
 1. Renovate searches in each repository for any `Cargo.toml` files
 1. Renovate extracts existing dependencies from `[dependencies]`, `[dev-dependencies]`, `[build-dependencies]` and `[workspace.dependencies]`
-1. Renovate tries to find and parse a `.cargo/config.toml` file to discover index URLs for private registries
+1. Renovate looks up Cargo configuration to discover index URLs for private registries.
```
Comment (HonkingGoose) [inline docs/usage/rust.md:14]:
```suggestion
1. Renovate looks up Cargo configuration to find index URLs for private registries
```

We don't use punctuation at the end of a list item. I like `find` better than `discover`.

File: docs/usage/rust.md [hunk 3]
```diff
@@ -11,14 +11,19 @@ Renovate supports upgrading dependencies in `Cargo.toml` files and their accompa
...
 1. Renovate resolves the dependency's version using the crates.io API or by cloning the index URL
 1. If Renovate finds an update, Renovate will use `cargo update` to update both `Cargo.toml` and `Cargo.lock`
 
 ## Enabling Rust Modules Updating
 
 Renovate updates Rust crates by default.
 
+## Cargo configuration and private registry discovery
+
+Renovate can determine private registry URLs from a `.cargo/config.toml` (or the legacy `.cargo/config`) Cargo configuration file.
```
Comment (HonkingGoose) [inline docs/usage/rust.md:24]:
```suggestion
Renovate can find private registry URLs in these Cargo configuration files:

- `.cargo/config.toml`
- `.cargo/config` (legacy)
```

Changes:

- Put config files in a list
- Replace `determine` with the simple word `find`
- Rewrite the sentence a little bit

File: docs/usage/rust.md [hunk 4]
```diff
@@ -11,14 +11,19 @@ Renovate supports upgrading dependencies in `Cargo.toml` files and their accompa
...
+Renovate can also detect private registry URLs via a `CARGO_REGISTRIES_<name>_INDEX` [environment variable](https://doc.rust-lang.org/cargo/reference/environment-variables.html#configuration-environment-variables).
```
Comment (HonkingGoose) [inline docs/usage/rust.md:25]:
```suggestion
Renovate can also find private registry URLs via a `CARGO_REGISTRIES_<name>_INDEX` environment variable.
Read the [Rust environment variables docs](https://doc.rust-lang.org/cargo/reference/environment-variables.html#configuration-environment-variables) to learn more.
```

How about this:

- replace `detect` with the simple word `find`
- Split the sentence into two parts
- Use a better name for the link to the Rust docs for the environment variable

**Review by HonkingGoose — CHANGES_REQUESTED** [review]

File: docs/usage/rust.md [hunk 5]
```diff
@@ -11,14 +11,23 @@ Renovate supports upgrading dependencies in `Cargo.toml` files and their accompa
...
 
 Renovate updates Rust crates by default.
 
+## Cargo configuration and private registry discovery
+
+Renovate can find private registry URLs in these Cargo configuration files:
+
+- `.cargo/config.toml`
+- `.cargo/config` (legacy)
+Renovate can also find private registry URLs via a `CARGO_REGISTRIES_<name>_INDEX` environment variable.
```
Comment (HonkingGoose) [inline docs/usage/rust.md:28]:
```suggestion

Renovate can also find private registry URLs via a `CARGO_REGISTRIES_<name>_INDEX` environment variable.
```

File: lib/modules/manager/cargo/extract.ts [hunk 6]
```diff
@@ -43,7 +53,15 @@ function extractFromSection(
...
         if (registryName) {
-          const registryUrl = cargoRegistries[registryName];
+          let registryUrl = cargoRegistries[registryName];
+
+          if (!registryUrl) {
+            const envUrl = get_cargo_index_env(registryName);
+            if (envUrl) {
+              registryUrl = envUrl;
+            }
+          }
```
Comment (secustor) [inline lib/modules/manager/cargo/extract.ts:63]:
```suggestion
          const registryUrl = cargoRegistries[registryName] ?? getCargoIndexEnv(registryName);
```
Same as above and nested if clause has no effect.

File: lib/modules/manager/cargo/extract.ts [hunk 7]
```diff
@@ -11,6 +11,16 @@ import type {
...
 
+function get_cargo_index_env(registryName: string): string | null {
+  return (
+    process.env[
+      `CARGO_REGISTRIES_${registryName
+        .toUpperCase()
+        .replaceAll('-', '_')}_INDEX`
+    ] ?? null
+  );
+}
```
Comment (secustor) [inline lib/modules/manager/cargo/extract.ts:22]:
```suggestion
function getCargoIndexEnv(registryName: string): string | null {
  const registry = registryName.toUpperCase().replaceAll('-', '_');
  return process.env[`CARGO_REGISTRIES_${registry}_INDEX`] ?? null;
}
```
Use camelCase and improve readability.

**Review by secustor — CHANGES_REQUESTED** [review]

**Review by secustor — DISMISSED** [review]

**Review by HonkingGoose — DISMISSED** [review]
I'm happy with the docs.

**Review by secustor — APPROVED** [review]

**Review by viceice — APPROVED** [review]

**Pull request merged**